### PR TITLE
Fix DDP on XLA

### PR DESCRIPTION
**Pull request by Liyang90**
Fix DDP on XLA by inserting `xm.reduce_gradients` in `closure` instead of using `xm.optimizer_step`. Adding a `xm.mark_step()` call after `optimizer.step` for performance.

## What does this PR do?

It fixes DDP correctness when using `tpu` accelerator. The related issue is [15878](https://github.com/Lightning-AI/lightning/issues/15878)

It also includes fix for "Issue 1" in [15884](https://github.com/Lightning-AI/lightning/issues/15884)

Fixes #\15878

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/pytorch_lightning/plugins/precision/tpu.py [hunk 1]
```diff
@@ -39,8 +46,10 @@ def optimizer_step(  # type: ignore[override]
     ) -> Any:
         import torch_xla.core.xla_model as xm
 
+        closure = partial(self._tpu_wrap_closure, optimizer, closure)
         closure = partial(self._wrap_closure, model, optimizer, optimizer_idx, closure)
-        closure_result = xm.optimizer_step(optimizer, optimizer_args={"closure": closure, **kwargs})
+        closure_result = optimizer.step(closure=closure, **kwargs)
+        xm.mark_step()
```
Comment (awaelchli) [inline src/pytorch_lightning/plugins/precision/tpu.py:52]: So essentially what we needed was barrier=True here, but to make it work with the closure, you had to pull out the implementation from around the closure: https://github.com/pytorch/xla/blob/b2bd721cf18b496feec02f12313bb532484eaab9/torch_xla/core/xla_model.py#L1030-L1032
Is my explanation correct?

**Review by awaelchli — APPROVED** [review]

File: src/pytorch_lightning/plugins/precision/tpu.py (hunk 1 above)
Comment (Liyang90) [reply src/pytorch_lightning/plugins/precision/tpu.py:52]: Yes, that's right.

**Review by Borda — CHANGES_REQUESTED** [review]
lgth, just can we add a test for this discovery?

**Comment by awaelchli** [comment]
@Borda I added the test.

**Review by Borda — APPROVED** [review]

**Review by justusschock — APPROVED** [review]
